core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/tool/JsonSchemaElementObjectSubstitution.java

```diff
@@ -0,0 +1,86 @@
+package io.quarkiverse.langchain4j.runtime.tool;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import dev.langchain4j.model.chat.request.json.JsonArraySchema;
+import dev.langchain4j.model.chat.request.json.JsonBooleanSchema;
+import dev.langchain4j.model.chat.request.json.JsonEnumSchema;
+import dev.langchain4j.model.chat.request.json.JsonIntegerSchema;
+import dev.langchain4j.model.chat.request.json.JsonNumberSchema;
+import dev.langchain4j.model.chat.request.json.JsonObjectSchema;
+import dev.langchain4j.model.chat.request.json.JsonReferenceSchema;
+import dev.langchain4j.model.chat.request.json.JsonSchemaElement;
+import dev.langchain4j.model.chat.request.json.JsonStringSchema;
+import io.quarkus.runtime.ObjectSubstitution;
+
+public sealed class JsonSchemaElementObjectSubstitution
+        implements ObjectSubstitution<JsonSchemaElement, JsonSchemaElementObjectSubstitution.Serialized>
+        permits JsonArraySchemaObjectSubstitution,
+        JsonBooleanSchemaObjectSubstitution,
+        JsonEnumSchemaObjectSubstitution,
+        JsonIntegerSchemaObjectSubstitution,
+        JsonNumberSchemaObjectSubstitution,
+        JsonObjectSchemaObjectSubstitution,
+        JsonReferenceSchemaObjectSubstitution,
+        JsonStringSchemaObjectSubstitution {
+
+    // Using ConcurrentHashMap in case multiple threads are using this class at the same time
+    // Not sure if this will ever happen
+    private final Map<Class<?>, JsonSchemaElementObjectSubstitution> substitutions = new ConcurrentHashMap<>(8);
```

Comment:
@geoand I'm not too familiar with substitutions/etc. I wasn't sure if there was any kind of a chance of multi-threading here, so I used a `ConcurrentHashMap`.

If there's no chance of multiple threads calling this at the same time I can switch it to a `HashMap`.

Reply:
There is no need for this complexity. I prefer to use what @andreadimaio has done

Reply:
I will fix this. One difference between what I and @andreadimaio did was that he removed all traces of `toolParameters` from existence within Quarkus LangChain4j.

Is that something we can/should do? Or leave it as deprecated for now? If we remove it then that is a breaking change to other people who may be using/extending from us.

Reply:
Good point. Let's just deprecate for now

Reply:
The general idea is that Quarkus needs to be able to take an object passed to a recorder and create bytecode instructions out of it. For POJOs (or stuff like records), it's straighforward (as it is for example when serializing to JSON). But for more complex types, an extension needs to essentially instruct Quarkus on how to serialize / deserialize these types

> I'm also not clear on what a substitution actually is.

It's a way to allow Quarkus serialize (and deserialize) a complex object where normal POJO rules don't apply. 

> Do I need an ObjectSubstitution class specific for the generic JsonSchemaElement interface? Or does having an ObjectSubstitution for every implementing class suffice?

The latter.

Reply:
> I will fix this. One difference between what I and @andreadimaio did was that he removed all traces of `toolParameters` from existence within Quarkus LangChain4j.
> 
> Is that something we can/should do? Or leave it as deprecated for now? If we remove it then that is a breaking change to other people who may be using/extending from us.

I removed all traces of `toolParameters` because it is not possible to use both the `parameters` and `toolParameters` fields simultaneously.
```java
private ToolSpecification(Builder builder) {
        this.name = builder.name;
        this.description = builder.description;
        if (builder.parameters != null && builder.toolParameters != null) {
            throw new IllegalArgumentException("Both (new) JsonObjectSchema and (old) ToolParameters " +
                    "are used to specify tool parameters. Please use only (new) JsonObjectSchema.");
        }
        this.parameters = builder.parameters;
        this.toolParameters = builder.toolParameters;
    }
```
Since the `ToolProcessor` class now exclusively uses the parameters field, the `toolParameters` field will always be null. As a result, I decided to remove all traces of `toolParameters`.